### DISCUSSION TITLE
Integrate with sqlite-zstd

### DISCUSSION THREAD

**Discussion by franchb** [ncruces/go-sqlite3#314] [Q&A] [answered]
Is it possible to add [sqlite-zstd](https://github.com/phiresky/sqlite-zstd) extension to `ncruces/go-sqlite3` WASM driver?

**Comment by ncruces** [comment, marked as answer]
Maybe, but it really depends on what you mean.

Any extension you can compile and statically link into your own Wasm, is probably fair game.

I can't give specific instructions for [sqlite-zstd](https://github.com/phiresky/sqlite-zstd), but you can check how `sqlite3.wasm` is [built](https://github.com/ncruces/go-sqlite3/blob/main/embed/build.sh), and how various extensions are [linked and initialized](https://github.com/ncruces/go-sqlite3/blob/main/sqlite3/main.c). Then it's just a matter of [using your own Wasm](https://pkg.go.dev/github.com/ncruces/go-sqlite3#pkg-variables).

If, however, want you really want is _readonly_ compressed databases, my suggestion is that you compress your entire database with [Zstandard Seekable Format](https://github.com/facebook/zstd/blob/dev/contrib/seekable_format/zstd_seekable_compression_format.md), and use [zstd-seekable-format-go](https://github.com/SaveTheRbtz/zstd-seekable-format-go) with the [reader VFS](https://pkg.go.dev/github.com/ncruces/go-sqlite3/vfs/readervfs).

**Comment by franchb** [comment]
Thank you! I'll give [zstd-seekable-format-go](https://github.com/SaveTheRbtz/zstd-seekable-format-go) a shot.

**Reply by ncruces** [reply]
Please report back!

**Comment by muety** [comment]
> [...] compress your entire database with [Zstandard Seekable Format](https://github.com/facebook/zstd/blob/dev/contrib/seekable_format/zstd_seekable_compression_format.md), and use [zstd-seekable-format-go](https://github.com/SaveTheRbtz/zstd-seekable-format-go) with the [reader VFS](https://pkg.go.dev/github.com/ncruces/go-sqlite3/vfs/readervfs).

Just for my understanding, this solution would only be suitable for _reading_ an already compressed database, but wouldn't support to read and write to the database from Go as usual, with compression happening transparently in the background?

Did you give it a try nevertheless, @franchb? 

Would love to have a compression VFS implementation!

Apart from that, thanks a lot for your efforts, @ncruces!

**Reply by ncruces** [reply]
Yes, `zstd-seekable-format-go` would be *strictly* read-only. It's hard to envision something else TBH.

[The architecture for the official ZIPVFS](https://www.sqlite.org/zipvfs/doc/trunk/www/howitworks.wiki) depends on non public APIs.

As detailed on that page, the SQLite pager depends on writing fixed size power-of-two pages to power-of-two offsets. I was able to circumvent the issue for my encryption extensions, by using encryption schemes that do not change the size of pages/sectors.

Changing the size of pages is kinda the point of compression. More importantly, a page can be really small when first written to (bunch of zeros for an half-filled page) then updated to be incompressible. We can't just make something in the middle of a file grow.

The way the ZIPVFS works is kinda like there's a database “in memory” that saves compressed pages as BLOBs to a *second* database on disk, just more optimized. But we have no special APIs (other than SQL) to make this fast. I wonder, at that point why not use some KV store.

Anyway, this is more work than I can take in. But hopefully making this a discussion gives others ideas to work on.
